### PR TITLE
feat: Added new method: check for new unlisted tags

### PR DESCRIPTION
**Pull request by airween**
Motivation: store actually used tags in a file and check PR's that any rules contains a new (unlisted) tag or not.

With this step, we can exclude the addition of unwanted new tags or the misspelling of tags and helps to keep tags in a consistent state.

### CODE REVIEW

File: util/crs-rules-check/README.md
Comment (theseion) [inline util/crs-rules-check/README.md:51]:
```suggestion
* **Check rule tags** - only tags listed in `TAGS_USED` may be used as tags in rules
```

File: util/crs-rules-check/README.md
Comment (theseion) [inline util/crs-rules-check/README.md:52]:
```suggestion
    * to use a new tag on a rule, it **must** first be registered in the TAGS_USED file
```

File: util/crs-rules-check/rules-check.py
Comment (theseion) [inline util/crs-rules-check/rules-check.py:575]:
```suggestion
        check that only tags from the TAGS_USED file are used
```

File: util/crs-rules-check/rules-check.py
Comment (theseion) [inline util/crs-rules-check/rules-check.py:600]:
```suggestion
                                'message': "rule uses unknown tag: '%s'; only tags registered in the TAGS_USED file may be used; rule id: %d" % (a['act_arg'], ruleid)
```

File: util/crs-rules-check/rules-check.py
Comment (theseion) [inline util/crs-rules-check/rules-check.py:695]:
```suggestion
    parser.add_argument("-t", "--tags-list", dest="tagslist", help="Path to file with permitted tags", required=True)
```

File: util/crs-rules-check/rules-check.py
Comment (theseion) [inline util/crs-rules-check/rules-check.py:712]:
```suggestion
            # remove empty items, if any
```

**Review by theseion — CHANGES_REQUESTED** [review]

**Review by theseion — DISMISSED** [review]

**Comment by airween** [comment]
Please do not merge this until other devs haven't seen. I'm curious about others opinion too.

**Review by fzipi — DISMISSED** [review]
I think we should name it `APPROVED_TAGS` instead of `USED_TAGS`.

**Comment by airween** [comment]
> I think we should name it `APPROVED_TAGS` instead of `USED_TAGS`.

Thanks. I wait with the modifications till other opinions.

**Comment by theseion** [comment]
I like the idea to use `APPROVED_TAGS`.

**Comment by dune73** [comment]
No strong feelings on the `APPROVED_TAGS` vs `USED_TAGS` question.

Maybe this config file should not be on the top level, though. It looks a bit out of place. Maybe in tests or util folder.

I like the supplementary `--tags-list` option.

Other than that I do like the general idea of the PR.

**Comment by airween** [comment]
> Maybe this config file should not be on the top level, though. It looks a bit out of place. Maybe in tests or util folder.

@fzipi, @theseion, @dune73 - please help me what would be the good place for the file of list of tags.

**Comment by dune73** [comment]
There are not that many options. Checking the landscape a bit, I think util would be right place. Right next to the file `id-range`.

**Comment by theseion** [comment]
> > Maybe this config file should not be on the top level, though. It looks a bit out of place. Maybe in tests or util folder.
> 
> @fzipi, @theseion, @dune73 - please help me what would be the good place for the file of list of tags.

For me there are three options, in order of preference:
- top-level
- `util/crs-rules-check`
- `rules`

I wouldn't put it into `rules` because all the files in rules currently pertain to rule content only, not metadata.
I would rather not put it into `util/crs-rules-check` because it's core data, not particularly associated with the rules check script.

**Comment by airween** [comment]
Thank you all. After thinking about it, may be the top-level isn't a good choice. I wouldn't put it beside the `rules-check` script, because I'm sure it doesn't belong that - it belongs to CRS itself.

I think the `utils` directory will be a goo place.

Adding the cli switch `--tags-list` is a good idea, but then it must be mandatory.

I'm going to rename the file to `APPROVED_TAGS`.

Note: the idea of this feature came from [here](https://github.com/coreruleset/coreruleset/pull/3347#discussion_r1377237481).

**Comment by airween** [comment]
Every requested modifications are done.

**Comment by dune73** [comment]
Given the `APPROVED_TAG` renaming was implemented like @fzipi asked for, I'm dismissing his review and merging. Thank you all.

**Pull request merged**
